### PR TITLE
CLN: Comparison methods for MultiIndex should have consistent behaviour for all nlevels  (GH21149)

### LINKED ISSUE #21149: (Assuming we don't want to disallow 1-level MultiIndexes), stop identifying MultiIndex by nlevels == 1
#### Code Sample, a copy-pastable example if possible

```python
In [2]: mi = pd.MultiIndex.from_product([[0, 1]])

In [3]: mi.set_names(('first',), level=0)
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-3-1f4ad6c7f0bf> in <module>()
----> 1 mi.set_names(('first',), level=0)

/home/nobackup/repo/pandas/pandas/core/indexes/base.py in set_names(self, names, level, inplace)
   1386 
   1387         if level is not None and self.nlevels == 1:
-> 1388             raise ValueError('Level must be None for non-MultiIndex')
   1389 
   1390         if level is not None and not is_list_like(level) and is_list_like(

ValueError: Level must be None for non-MultiIndex

```

#### Problem description

The case of ``MultiIndex`` with ``self.nlevels == 1`` is currently perfectly valid (despite e.g. ``MultiIndex.droplevel`` returning a normal ``Index`` when only one level is left). Unless we want to change this, we should always identify a ``MultiIndex`` with ``isinstance``, not by looking at ``self.nlevels``.

#### Expected Output

```python
In [4]: mi.set_names(('first',))
Out[4]: 
MultiIndex(levels=[[0, 1]],
           labels=[[0, 1]],
           names=['first'])
```

#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit: 5cf4957a2cc8429170742dd1e6435ddb8c323811
python: 3.5.3.final.0
python-bits: 64
OS: Linux
OS-release: 4.9.0-6-amd64
machine: x86_64
processor: 
byteorder: little
LC_ALL: None
LANG: it_IT.UTF-8
LOCALE: it_IT.UTF-8

pandas: 0.24.0.dev0+14.g5cf4957a2.dirty
pytest: 3.5.0
pip: 9.0.1
setuptools: 39.0.1
Cython: 0.25.2
numpy: 1.14.3
scipy: 0.19.0
pyarrow: None
xarray: None
IPython: 6.2.1
sphinx: 1.5.6
patsy: 0.5.0
dateutil: 2.7.3
pytz: 2018.4
blosc: None
bottleneck: 1.2.0dev
tables: 3.3.0
numexpr: 2.6.1
feather: 0.3.1
matplotlib: 2.0.0
openpyxl: 2.3.0
xlrd: 1.0.0
xlwt: 1.3.0
xlsxwriter: 0.9.6
lxml: 4.1.1
bs4: 4.5.3
html5lib: 0.999999999
sqlalchemy: 1.0.15
pymysql: None
psycopg2: None
jinja2: 2.10
s3fs: None
fastparquet: None
pandas_gbq: None
pandas_datareader: 0.2.1

</details>

### PR DESCRIPTION
**Pull request by KalyanGokhale**
- [x] closes #21149 
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

Initial PR #21182 is closed - and now split in 2 different PRs

`def cmp_method` raised ValueError for equality / inequality comparisons of MultiIndex with nlevels == 1, which was inconsistent with behaviour for MultiIndex with nlevels > 1 (details of issue below) - this has now been fixed

Currently (as of 0.23.0), comparing MultiIndex of nlevels==1 with another of same length raises a ValueError e.g.
```python
[In] midx=pd.MultiIndex.from_product([[0, 1]])
[In] midx
[Out] MultiIndex(levels=[[0, 1]],
           labels=[[0, 1]])
[In] midx == midx
[Out] ValueError: The truth value of an array with more than one element is ambiguous. Use a.any() or a.all()
```
whereas the behaviour should be consistent with that for MultiIndex with nlevels>1 as follows:
```python
[In] midx == midx
[Out] array([ True,  True])
```

### CODE REVIEW

File: doc/source/whatsnew/v0.23.1.txt [hunk 1]
```diff
@@ -78,7 +78,8 @@ Indexing
 
 - Bug in :meth:`Series.reset_index` where appropriate error was not raised with an invalid level name (:issue:`20925`)
 - Bug in :func:`interval_range` when ``start``/``periods`` or ``end``/``periods`` are specified with float ``start`` or ``end`` (:issue:`21161`)
--
+- Bug in comparison operations for :class:`MultiIndex` where error was raised on equality / inequality comparison involving a MultiIndex with self.nlevels == 1 (:issue:`21149`)
+-
```
Comment (jreback) [inline doc/source/whatsnew/v0.23.1.txt]: double backticks on MultiIndex

File: pandas/tests/indexes/test_multi.py [hunk 2]
```diff
@@ -3291,3 +3291,29 @@ def test_duplicate_multiindex_labels(self):
         with pytest.raises(ValueError):
             ind.set_levels([['A', 'B', 'A', 'A', 'B'], [2, 1, 3, -2, 5]],
                            inplace=True)
+
+    @pytest.mark.parametrize("midx,idx,count", [
+        (pd.MultiIndex.from_product([[0, 1], [1, 0]]), pd.Series(range(4)), 4),
+        (pd.MultiIndex.from_product([[0, 1]]), pd.Series(range(2)), 2)])
+    def test_multiindex_compare(self, midx, idx, count):
+        # GH 21149
+        '''Ensure comparison operations for MultiIndex with nlevels == 1
```
Comment (jreback) [inline pandas/tests/indexes/test_multi.py]: use triple-double quotes

File: pandas/tests/indexes/test_multi.py [hunk 3]
```diff
@@ -3291,3 +3291,29 @@ def test_duplicate_multiindex_labels(self):
...
+            behave consistently with those for MultiIndex with nlevels > 1
+        '''
+        expected = pd.Series([True]).repeat(count)
+        expected.reset_index(drop=True, inplace=True)
+        # Equality self-test: MultiIndex object vs self
```
Comment (jreback) [inline pandas/tests/indexes/test_multi.py]: blank line between cases

File: pandas/tests/indexes/test_multi.py [hunk 4]
```diff
@@ -3291,3 +3291,29 @@ def test_duplicate_multiindex_labels(self):
...
+    @pytest.mark.parametrize("midx,idx,count", [
+        (pd.MultiIndex.from_product([[0, 1], [1, 0]]), pd.Series(range(4)), 4),
+        (pd.MultiIndex.from_product([[0, 1]]), pd.Series(range(2)), 2)])
+    def test_multiindex_compare(self, midx, idx, count):
+        # GH 21149
+        '''Ensure comparison operations for MultiIndex with nlevels == 1
+            behave consistently with those for MultiIndex with nlevels > 1
+        '''
+        expected = pd.Series([True]).repeat(count)
+        expected.reset_index(drop=True, inplace=True)
```
Comment (jreback) [inline pandas/tests/indexes/test_multi.py]: don't use inplace in tests

File: pandas/tests/indexes/test_multi.py [hunk 5]
```diff
@@ -3291,3 +3291,29 @@ def test_duplicate_multiindex_labels(self):
...
+        # Equality self-test: MultiIndex object vs self
+        result = pd.Series(midx == midx)
+        tm.assert_series_equal(result, expected)
+        # Equality self-test: non-MultiIndex Index object vs self
+        result = (idx == idx)
```
Comment (jreback) [inline pandas/tests/indexes/test_multi.py]: where did the idea for these test come from? what exactly are you trying to test here?

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by toobaz** [comment]
> def cmp_method raised ValueError for equality / inequality comparisons of MultiIndex with nlevels == 1

Good catch!

File: pandas/tests/indexes/test_multi.py (hunk 4 above)
Comment (KalyanGokhale) [reply pandas/tests/indexes/test_multi.py]: @jreback Thanks - here `inplace` is being used to only create the expected outcome for testing. `inplace` is not being used in the test `assert_series_equal`
OK to keep then?

File: pandas/tests/indexes/test_multi.py (hunk 5 above)
Comment (KalyanGokhale) [reply pandas/tests/indexes/test_multi.py]: These tests are trying to ensure the behaviour on comparison between `MultiIndex vs MultiIndex`, `MultiIndex vs Index` is consistent, irrespective of `nlevels`. Also, ensuring that the comparison behaviour for `Index vs Index` has not changed due to change in this PR.

Currently (as of 0.23.0), comparing MultiIndex of nlevel==1 with another of same length raises a ValueError e.g.
```python
[In] midx=pd.MultiIndex.from_product([[0, 1]])
[In] midx
[Out] MultiIndex(levels=[[0, 1]],
           labels=[[0, 1]])
[In] midx == midx
[Out] ValueError: The truth value of an array with more than one element is ambiguous. Use a.any() or a.all()
```

whereas the behaviour should be consistent with that for MultiIndex with nlevel>1 as follows:
```python
[In] midx == midx
[Out] array([ True,  True])
```

Comment (KalyanGokhale) [reply pandas/tests/indexes/test_multi.py]: These tests are trying to ensure the behaviour on comparison between `MultiIndex vs MultiIndex`, `MultiIndex vs Index` is consistent, irrespective of `nlevels`. Also, ensuring that the comparison behaviour for `Index vs Index` has not changed due to change in this PR.

Currently (as of 0.23.0), comparing MultiIndex of nlevel==1 with another of same length raises a ValueError e.g.
```python
[In] midx=pd.MultiIndex.from_product([[0, 1]])
[In] midx
[Out] MultiIndex(levels=[[0, 1]],
           labels=[[0, 1]])
[In] midx == midx
[Out] ValueError: The truth value of an array with more than one element is ambiguous. Use a.any() or a.all()
```
whereas the behaviour should be consistent with that for MultiIndex with nlevel>1 as follows:
```python
[In] midx == midx
[Out] array([ True,  True])
```

Comment (KalyanGokhale) [reply pandas/tests/indexes/test_multi.py]: in this test, probably should be OK to only keep the `midx == midx` and `midx > midx` checks

**Comment by KalyanGokhale** [comment]
@jreback @toobaz are any other edits needed on this? Thanks

**Comment by KalyanGokhale** [comment]
Any other work needed on this PR? Thanks

**Review by jreback — APPROVED** [review]

**Pull request merged**